resourcemanager/commonids/app_service_plan.go

```diff
@@ -0,0 +1,127 @@
+type AppServicePlanId struct {
+	SubscriptionId    string
+	ResourceGroupName string
+	ServerFarmName    string
+}
+
+// NewAppServicePlanID returns a new AppServicePlanId struct
+func NewAppServicePlanID(subscriptionId string, resourceGroupName string, serverFarmName string) AppServicePlanId {
+	return AppServicePlanId{
+		SubscriptionId:    subscriptionId,
+		ResourceGroupName: resourceGroupName,
+		ServerFarmName:    serverFarmName,
+	}
+}
+
+// ParseAppServicePlanID parses 'input' into a AppServicePlanId
+func ParseAppServicePlanID(input string) (*AppServicePlanId, error) {
+	parser := resourceids.NewParserFromResourceIdType(AppServicePlanId{})
+	parsed, err := parser.Parse(input, false)
+	if err != nil {
+		return nil, fmt.Errorf("parsing %q: %+v", input, err)
+	}
+
+	var ok bool
+	id := AppServicePlanId{}
+
+	if id.SubscriptionId, ok = parsed.Parsed["subscriptionId"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "subscriptionId", *parsed)
+	}
+
+	if id.ResourceGroupName, ok = parsed.Parsed["resourceGroupName"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "resourceGroupName", *parsed)
+	}
+
+	if id.ServerFarmName, ok = parsed.Parsed["serverFarmName"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "serverFarmName", *parsed)
+	}
+
+	return &id, nil
+}
+
+// ParseAppServicePlanIDInsensitively parses 'input' case-insensitively into a AppServicePlanId
+// note: this method should only be used for API response data and not user input
+func ParseAppServicePlanIDInsensitively(input string) (*AppServicePlanId, error) {
+	parser := resourceids.NewParserFromResourceIdType(AppServicePlanId{})
+	parsed, err := parser.Parse(input, true)
+	if err != nil {
+		return nil, fmt.Errorf("parsing %q: %+v", input, err)
+	}
+
+	var ok bool
+	id := AppServicePlanId{}
+
+	if id.SubscriptionId, ok = parsed.Parsed["subscriptionId"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "subscriptionId", *parsed)
+	}
+
+	if id.ResourceGroupName, ok = parsed.Parsed["resourceGroupName"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "resourceGroupName", *parsed)
+	}
+
+	if id.ServerFarmName, ok = parsed.Parsed["serverFarmName"]; !ok {
+		return nil, resourceids.NewSegmentNotSpecifiedError(id, "serverFarmName", *parsed)
+	}
+
+	return &id, nil
+}
+
+// ValidateAppServicePlanID checks that 'input' can be parsed as an App Service Plan ID
+func ValidateAppServicePlanID(input interface{}, key string) (warnings []string, errors []error) {
+	v, ok := input.(string)
+	if !ok {
+		errors = append(errors, fmt.Errorf("expected %q to be a string", key))
+		return
+	}
+
+	if _, err := ParseAppServicePlanID(v); err != nil {
+		errors = append(errors, err)
+	}
+
+	return
+}
+
+// ID returns the formatted App Service Plan ID
+func (id AppServicePlanId) ID() string {
+	fmtString := "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Web/serverFarms/%s"
```

Comment:
Discussed offline at length, but calling out again here for completeness. Since `serverfarm` is considered "one word" in the service, this will need a state migration and careful testing as some older aspects of the API may be case sensitive.

Reply:
Yeah indeed - the Portal is using `serverFarms` at this point in time, which whilst I'm aware isn't a definitive source (and the Swagger is) - given the ARM conventions around resource naming (that we should be using camelCase) I think it's worth us running with that in this instance - but yeah, we _are_ going to need state migrations (and to parse the IDs insensitively in the Read functions etc)

Reply:
(chatting offline, this is fine)